### PR TITLE
Manage block device Space.

### PR DESCRIPTION
**Pull request by HHoflittlefish777**
Work about this pr:
-  Manage block space with a bit
-  Manage slice space with lock free skip list.

future tasks:
-  verify correct.
-  benchmark

### CODE REVIEW

**Review by luanshaotong — CHANGES_REQUESTED** [review]
Your PR needs more complete implement.

**Comment by uran0sH** [comment]
I think you should propose an RFC first.

**Comment by HHoflittlefish777** [comment]
> I think you should propose an RFC first.

So, it is become the first RFC?

**Comment by luanshaotong** [comment]
I think the main issue with this PR is completeness. At the same time, it is difficult for others to continue to work on the basis of this PR, if the author gives up. So please submit a standalone module that does not require subsequent implementation.
The RFC does not affect this issue. Of course, it would be great to have an RFC.

**Comment by uran0sH** [comment]
Why not use crossbeam-skiplist directly? Do you think your skiplist performance is better than it?

File: src/server/storage_engine/block_device/block_storage.rs [hunk 1]
```diff
@@ -2,6 +2,20 @@
...
+
+use super::allocator::SkipListAllocator;
+
+//#define BLKGETSIZE _IO(0x12,96)	/* return device size /512 (long *arg) */
+const _BLOCKGETSIZE: u64 = 0x1260;
+
+const _CHUNK: u64 = 512 * 8;
+const _SECTOR: u64 = 512;
+
 pub trait BlockStorage {
```
Comment (uran0sH) [inline src/server/storage_engine/block_device/block_storage.rs:19]: Why need this trait?

**Review by uran0sH — CHANGES_REQUESTED** [review]

**Comment by HHoflittlefish777** [comment]
> Why not use crossbeam-skiplist directly? Do you think your skiplist performance is better than it?

There are two reasons. First, the library has not been updated for five years. Some places are incompatible with the new one. Or if there is a problem, who can solve it? Second, in our scenario, we need to merge data blocks of the same kind, so from this perspective, I think it may not be suitable

**Comment by uran0sH** [comment]
So what's difference between your skiplist and comman skiplist?
> Second, in our scenario, we need to merge data blocks of the same kind, so from this perspective,

**Comment by uran0sH** [comment]
https://github.com/crossbeam-rs/crossbeam/tree/master/crossbeam-skiplist Are you sure this hasn't been updated in five years?
> First, the library has not been updated for five years

**Comment by HHoflittlefish777** [comment]
> https://github.com/crossbeam-rs/crossbeam/tree/master/crossbeam-skiplist Are you sure this hasn't been updated in five years?
> 
> > First, the library has not been updated for five years

It is my mistake.

**Comment by luanshaotong** [comment]
I think you need to implement call route from storage_engine to your block interface, even it is a demo or mock.

**Pull request closed without merging**